File: server/app/schemas/garden.py
from pydantic import BaseModel
from typing import List, Optional, Union, Dict, Any
from datetime import datetime


# Garden Element Schemas
class GardenElementBase(BaseModel):
    element_id: str
    element_type: str  # 'structure', 'plant', 'text'
    position_x: float
    position_y: float

    # Structure fields
    width: Optional[float] = None
    height: Optional[float] = None
    label: Optional[str] = None
    color: Optional[str] = None
    shape: Optional[str] = None

    # Plant fields
    common_name: Optional[str] = None
    botanical_name: Optional[str] = None
    plant_type: Optional[str] = None
    sunlight_needs: Optional[str] = None
    water_needs: Optional[str] = None
    mature_size: Optional[str] = None
    spacing: Optional[float] = None
    show_spacing: Optional[bool] = False

    # Text fields
    text_content: Optional[str] = None
    font_size: Optional[int] = None
    text_color: Optional[str] = None


class GardenElementCreate(GardenElementBase):
    pass


class GardenElementUpdate(BaseModel):
    element_type: Optional[str] = None
    position_x: Optional[float] = None
    position_y: Optional[float] = None
    width: Optional[float] = None
    height: Optional[float] = None
    label: Optional[str] = None
    color: Optional[str] = None
    shape: Optional[str] = None
    common_name: Optional[str] = None
    botanical_name: Optional[str] = None
    plant_type: Optional[str] = None
    sunlight_needs: Optional[str] = None
    water_needs: Optional[str] = None
    mature_size: Optional[str] = None
    spacing: Optional[float] = None
    show_spacing: Optional[bool] = None
    text_content: Optional[str] = None
    font_size: Optional[int] = None
    text_color: Optional[str] = None


class GardenElement(GardenElementBase):
    id: int
    garden_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Garden Schemas
class GardenBase(BaseModel):
    name: str
    description: Optional[str] = None
    zip_code: str
    view_box_x: float = -500.0
    view_box_y: float = -500.0
    view_box_width: float = 1000.0
    view_box_height: float = 1000.0
    zoom: float = 1.0
    grid_size: int = 50


class GardenCreate(GardenBase):
    pass


class GardenUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    zip_code: Optional[str] = None
    view_box_x: Optional[float] = None
    view_box_y: Optional[float] = None
    view_box_width: Optional[float] = None
    view_box_height: Optional[float] = None
    zoom: Optional[float] = None
    grid_size: Optional[int] = None


class Garden(GardenBase):
    id: int
    user_id: str
    created_at: datetime
    updated_at: Optional[datetime] = None
    elements: List[GardenElement] = []
    notes: List["GardenNote"] = []

    class Config:
        from_attributes = True


class GardenSummary(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    zip_code: str
    created_at: datetime
    updated_at: Optional[datetime] = None
    element_count: int

    class Config:
        from_attributes = True


# Garden Note Schemas
class GardenNoteBase(BaseModel):
    content: str


class GardenNoteCreate(GardenNoteBase):
    pass


class GardenNote(GardenNoteBase):
    id: int
    garden_id: int
    created_at: datetime

    class Config:
        from_attributes = True


# Bulk operations
class GardenElementsBulkUpdate(BaseModel):
    elements: List[Union[GardenElementCreate, GardenElementUpdate]]


class GardenSnapshot(BaseModel):
    garden: GardenUpdate
    elements: List[GardenElementCreate]


# Recommendation Schemas
class GardenRecommendationBase(BaseModel):
    data: Dict[str, Any]


class GardenRecommendation(GardenRecommendationBase):
    id: int
    garden_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: client/src/types/garden.ts
export interface Plant {
  commonName: string;
  botanicalName: string;
  plantType: 'Shade Trees' | 'Fruit Trees' | 'Flowering Shrubs' | 'Vegetables' | 'Herbs';
  sunlightNeeds: string;
  waterNeeds: string;
  matureSize: string;
  spacing: number;
}

export interface PlantRecommendations {
  shadeTrees: Plant[];
  fruitTrees: Plant[];
  floweringShrubs: Plant[];
  vegetables: Plant[];
  herbs: Plant[];
}

export interface GeminiResponse {
  recommendedPlants: PlantRecommendations;
}

export interface Position {
  x: number;
  y: number;
}

export interface Size {
  width: number;
  height: number;
}

export interface CanvasObject {
  id: string;
  position: Position;
  type: 'structure' | 'plant' | 'text';
}

export type StructureShape = 'rectangle' | 'ellipse';

export interface Structure extends CanvasObject {
  type: 'structure';
  size: Size;
  label: string;
  color: string;
  shape: StructureShape;
  // Vertical height in feet for 3D mode extrusion
  zHeight?: number;
}

export interface PlantInstance extends CanvasObject {
  type: 'plant';
  plant: Plant;
  showSpacing: boolean;
}

export interface TextLabel extends CanvasObject {
  type: 'text';
  text: string;
  fontSize: number;
  color: string;
}

export type CanvasElement = Structure | PlantInstance | TextLabel;

export interface CanvasState {
  elements: CanvasElement[];
  selectedElementId: string | null;
  viewBox: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
  zoom: number;
  gridSize: number;
}

export type Tool = 'select' | 'structure' | 'text';

export interface GardenNote {
  id: number;
  garden_id: number;
  content: string;
  created_at: string;
}

// Garden persistence types
export interface SavedGarden {
  id: number;
  name: string;
  description?: string;
  zip_code: string;
  user_id: string;
  view_box_x: number;
  view_box_y: number;
  view_box_width: number;
  view_box_height: number;
  zoom: number;
  grid_size: number;
  created_at: string;
  updated_at?: string;
  elements: SavedGardenElement[];
}

export interface SavedGardenElement {
  id: number;
  element_id: string;
  garden_id: number;
  element_type: string;
  position_x: number;
  position_y: number;
  width?: number;
  height?: number;
  z_height?: number;
  label?: string;
  color?: string;
   shape?: string;
  common_name?: string;
  botanical_name?: string;
  plant_type?: string;
  sunlight_needs?: string;
  water_needs?: string;
  mature_size?: string;
  spacing?: number;
  show_spacing?: boolean;
  text_content?: string;
  font_size?: number;
  text_color?: string;
  created_at: string;
  updated_at?: string;
}

export interface GardenSummary {
  id: number;
  name: string;
  description?: string;
  zip_code: string;
  created_at: string;
  updated_at?: string;
  element_count: number;
}

export interface AppState {
  canvas: CanvasState;
  activeTool: Tool;
  structureShape: StructureShape;
  plantRecommendations: PlantRecommendations | null;
  userZipCode: string | null;
  isLoadingRecommendations: boolean;
  sidePanel: 'recommendations' | 'properties' | 'gardens' | 'notes';
  currentGarden: SavedGarden | null;
  notes: GardenNote[];
  isSaving: boolean;
  lastSaved: Date | null;
  isLoadingGardens: boolean;
  availableGardens: GardenSummary[];
  showPlantSpacing: boolean;
  // Whether to render in 3D mode
  is3DMode: boolean;
}